### PR TITLE
Add TemplateByID, TemplateVersion, Templates and TemplatePreview

### PR DESCRIPTION
**Pull request by andrewoh531**
…o add integration tests.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -69,12 +70,50 @@ func NewClient(apiKey string, options ...ClientOption) (*Client, error) {
 	return &Client{client}, nil
 }
 
-func (c Client) Templates() (Templates, error) {
+func (c Client) TemplateByID(templateID string) (Template, error) {
```
Comment (jonathaningram) [inline client.go]: @andrewoh531 can just use `id` as argument name.

File: client.go [hunk 2]
```diff
@@ -69,12 +70,50 @@ func NewClient(apiKey string, options ...ClientOption) (*Client, error) {
...
+	var template Template
+	err := c.c.Get("./v2/template/" + templateID).JSON(&template).Error
+	return template, err
+}
+
+func (c Client) TemplateVersion(templateID string, version int) (Template, error) {
```
Comment (jonathaningram) [inline client.go]: @andrewoh531 can just use `id` as argument name.

File: client.go [hunk 3]
```diff
@@ -69,12 +70,50 @@ func NewClient(apiKey string, options ...ClientOption) (*Client, error) {
...
+	url := "/v2/template/" + templateID + "/version/" + strconv.Itoa(version)
+	var template Template
+	err := c.c.Get(url).JSON(&template).Error
+	return template, err
+}
+
+func (c Client) Templates(templateType string) (Templates, error) {
```
Comment (jonathaningram) [inline client.go]: @andrewoh531 can just use `typ` as argument name (is quite idiomatic to name `type` as `typ`).

File: client.go [hunk 4]
```diff
@@ -69,12 +70,50 @@ func NewClient(apiKey string, options ...ClientOption) (*Client, error) {
...
+	}
+
 	var templates Templates
-	err := c.c.Get("./v2/templates").JSON(&templates, "templates").Error
+	err := c.c.Get(url).JSON(&templates, "templates").Error
 	return templates, err
 }
 
+func (c Client) TemplatePreview(templateID string, personalisation ...PersonalisationOption) (TemplatePreview, error) {
+
```
Comment (jonathaningram) [inline client.go]: @andrewoh531 `templateID` to `id`, and remove newline after method signature.

File: client.go [hunk 5]
```diff
@@ -69,12 +70,50 @@ func NewClient(apiKey string, options ...ClientOption) (*Client, error) {
...
+	var response TemplatePreview
+	var buf bytes.Buffer
+	var payload = payload{}
```
Comment (jonathaningram) [inline client.go]: @andrewoh531 `var payload payload`

File: client_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		t.Fatalf("Error fetching email templates: %v", err)
+	}
+
+	assert.NotNil(t, resp)
+	assert.Equal(t, "Hi KD,\n\nMy favourite colour is yellow.\n", resp.Body)
+	assert.Equal(t, "email", resp.Type)
+}
+
+func assertTemplateFound(t *testing.T, templates notify.Templates, templateIds []string) {
+
```
Comment (jonathaningram) [inline client_integration_test.go]: Remove newline.

File: options.go [hunk 7]
```diff
@@ -69,12 +71,9 @@ func SMSSenderID(senderID string) SendSMSOption {
 	})
 }
 
-type payloadUpdater interface {
-	updatePayload(payload) payload
-}
-
-type CommonOption struct {
-	payloadUpdater
+type CommonOption interface {
```
Comment (jonathaningram) [inline options.go:74]: @andrewoh531 it's not clear why this changed? `CommonOption` now appears to be a publicly exported interface that contains two very specific interfaces.

File: client_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+import (
+	"os"
+	"testing"
+
+	notify "github.com/govau/notify-client-go"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var apiKey string
```
Comment (jonathaningram) [inline client_integration_test.go]: I'd prefer not to have global variables. Is there another way?

File: client_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+		client, err = notify.NewClient(
+			apiKey,
+		)
+		if err != nil {
+			t.Fatal("Error creating client", err)
+		}
+	}
+}
+
+func assertEqual(t *testing.T, expected interface{}, actual interface{}) {
```
Comment (jonathaningram) [inline client_integration_test.go]: Doesn't seem to be used?

File: client_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+	assert.NotNil(t, resp)
+	assertTemplateFound(t, resp, []string{smsTemplateId})
+}
+
+func TestGetEmailTemplates(t *testing.T) {
+	setup(t)
+	resp, err := client.Templates("email")
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: If you're going to use `testify` should this be `assert.NoError(t, err)`?

File: client_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+var apiKey string
+var smsTemplateId string
+var emailTemplateId string
+var phoneNumber string
+var emailAddress string
+var client *notify.Client
+
+func setup(t *testing.T) {
+	if client == nil {
```
Comment (jonathaningram) [inline client_integration_test.go]: If we keep this function, then return early if non-nil.

File: client_integration_test.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+)
+
+var apiKey string
+var smsTemplateId string
+var emailTemplateId string
+var phoneNumber string
+var emailAddress string
+var client *notify.Client
+
+func setup(t *testing.T) {
```
Comment (jonathaningram) [inline client_integration_test.go]: At the very least I think setup can return a `client` rather than setting up a global one. Should be reset every test anyway I think.

File: options.go (hunk 7 above)
Comment (andrewoh531) [reply options.go:74]: I might need to lean on @dagoof here as he helped me simplify this. I think the changes were to reduce the complexity.

Comment (jonathaningram) [reply options.go:74]: @dagoof you happy here?

Comment (dagoof) [reply options.go:74]: I'm happier with this than how it was; but I think there's still issues with what's going on here.

It's not exposed that `Personalisation` satisfies the `CommonOption` interface because we don't know what methods it takes to support the interface. Similarly, it's not totally clear that a `CommonOption` can be used as an argument to `SendSMS` which accepts `SendSMSOptions`s.

And even then, it's starting to seem unlikely that this approach scales beyond one or two functions. Do we need to create a `GetTemplateOption` interface, `GenerateTemplatePreviewOption` interface? for all methods that accept optional parameters.

We'll need to mint a new option interface (or multiple) for every exposed function; as well as constructors for all possible options for those interfaces. We'll end up with something like this for every single function:

```
type GetAllTemplatesOption interface { ... }

func TemplateType(string) GetAllTemplatesOption { ... }

// ... and more if other options are available

func GetAllTemplates(...GetAllTemplatesOption) TemplateList { ... }
```


I'm really not sure on the tradeoff. Sure we can very specifically describe what's available to each function in a type-safe and semantic way without any ambiguity. It takes a bit more work on our end but I don't mind that.

I'm just not sure if it ends up being any clearer to the user about how to use the thing over a plain `type GetAllTemplatesOptions struct { TemplateType string }` option struct

File: client_integration_test.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
+package notify_test
+
+import (
+	"os"
+	"testing"
+
+	notify "github.com/govau/notify-client-go"
+)
+
+func setup(t *testing.T) (*notify.Client, string, string) {
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 this is a good use case for named return arguments:

```go
func setup(t *testing.T) (client *notify.Client, smsTemplateID string, emailTemplateID string) {}
```

Because at a glance it's hard to see what the two `string, string`s are.

Also could you please change `xxxId` -> `xxxID` in the local variable names? Thanks!

File: client_integration_test.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	if emailTemplateId == "" {
+		t.Fatal("EMAIL_TEMPLATE_ID environment variable not set")
+	}
+
+	var err error
+	client, err := notify.NewClient(
+		apiKey,
+	)
+	if err != nil {
+		t.Fatal("Error creating client", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 probably better:`t.Fatalf("Could not create client: %v", err)`? `Could not` is a good convention for reporting errors.

File: client_integration_test.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+	client, _, emailTemplateId := setup(t)
+	resp, err := client.GetAllTemplates("email")
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
+	}
+
+	assertTemplateFound(t, resp, []string{emailTemplateId})
+}
+
+func TestGetTemplateById(t *testing.T) {
```
Comment (jonathaningram) [inline client_integration_test.go]: `Id` -> `ID`

File: client_integration_test.go (hunk 14 above)
Comment (jonathaningram) [reply client_integration_test.go]: @andrewoh531 Actually, per incoming suggestions use lowercase, so `could not...`

File: client_integration_test.go [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+		emailTemplateId,
+		emailAddress,
+		notify.Personalisation{
+			{"name", "John"},
+			{"colour", "pink"},
+		},
+		notify.Reference(ref),
+	)
+	if err != nil {
+		t.Fatalf("Error sending email: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: `could not send email`

File: client_integration_test.go [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+	})
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
+	}
+
+	assertEqual(t, "Hi KD,\n\nMy favourite colour is yellow.\n", resp.Body)
+	assertEqual(t, "email", resp.Type)
+}
+
+func assertEqual(t *testing.T, expected, actual interface{}) {
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 don't really even need to use this helper method, but if you're going to, more idiomatic is to reverse the argument order, use got and want and error message as follows:

```go
func assertEqual(t *testing.T, got, want interface{}) {
	if got != want {
		t.Errorf("got %v, want %v", got, want)
	}
}
```

The reason I'd not use it is because this for example:

```go
assertEqual(t, "email", resp.Type)
```

Would usually just be:

```go
want := "email" // And if this was a test from a table test, it would come off the table test, e.g. `tt.want` and so the local variable wouldn't be required.
if got := resp.Type; got != want {
	t.Errorf("Type = %v, want %v", got, want)
}
```

And now for a failed test, rather than reading generically, `got sms, want email` (or `Expected email but received sms`), it would read `Type = sms, want email` and so you now know what actually failed.

So in summary, after writing that, it's probably better to do this all inline and don't use the helper. Also the helper is really only a helper for a few types—it won't work if you try and pass it maps and slices, etc. In those cases, `reflect.DeepEqual()` is idiomatically used. With that in mind, `assertEqual` is a potential place for causing weird bugs in the tests.

File: go.mod [hunk 18]
```diff
@@ -3,6 +3,7 @@ module github.com/govau/notify-client-go
 go 1.12
 
 require (
+	github.com/stretchr/testify v1.3.0
```
Comment (jonathaningram) [inline go.mod]: @andrewoh531 you need to revert this.

File: go.sum [hunk 19]
```diff
@@ -1,3 +1,10 @@
+github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8=
```
Comment (jonathaningram) [inline go.sum]: @andrewoh531 you forgot to revert this.

File: options.go (hunk 7 above)
Comment (jonathaningram) [reply options.go:74]: Yeah not too sure without sitting on it for a bit and trying out some things. Are you both happy with the addition of an exported `PersonalisationOption` interface that we'll add with this PR?

Comment (andrewoh531) [reply options.go:74]: I'll defer to @dagoof

File: client_integration_test.go [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+	"os"
+	"testing"
+
+	notify "github.com/govau/notify-client-go"
+)
+
+func setup(t *testing.T) (client *notify.Client, smsTemplateID string, emailTemplateID string) {
+	apiKey := os.Getenv("API_KEY")
+	if apiKey == "" {
+		t.Fatal("could not set API_KEY environment variable")
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 message doesn't make sense. I think you meant `could not get API_KEY environment variable`. Also, since this is not really an error, rather a missing param, maybe `API_KEY environment variable must be set`. Splitting hairs between those two choices, either are fine.

File: client_integration_test.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+		smsTemplateID,
+		phoneNumber,
+		notify.Personalisation{
+			{"name", "John"},
+			{"day", "Friday"},
+		},
+		notify.Reference(ref),
+	)
+	if err != nil {
+		t.Errorf("could not send SMS: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 you do probably want to keep this as `t.Fatalf` so that the test exists here. Generally speaking you want the test to exit whenever there's an error, unless your test is concerned with checking what the error was and wanting to continue with other checks.

File: client_integration_test.go [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+	}
+	if resp.ID == "" {
+		t.Errorf("Response ID should not be empty")
+	}
+	if resp.URI == "" {
+		t.Errorf("Response URI should not be empty")
+	}
+	if *resp.Reference != ref {
+		t.Errorf("Got %v, want %v", *resp.Reference, ref)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 lowercase `G` 😎

File: client_integration_test.go [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+		apiKey,
+	)
+	if err != nil {
+		t.Fatal("could not create Notify client", err)
+	}
+	return
+}
+
+func TestSendSMS(t *testing.T) {
+	client, smsTemplateID, _ := setup(t)
```
Comment (jonathaningram) [inline client_integration_test.go:37]: @andrewoh531 see https://github.com/govau/cg-dashboard/commit/7989cc94f18bb85789e9b5f6002973389b9e7336

I think we should consider doing a similar thing e.g. `SKIP_INTEGRATION_TESTS`. Reason being that I checked this out locally and didn't have any env vars set up to successfully run all these tests. Something to consider, I think it might be a commonly used pattern.

Comment (andrewoh531) [reply client_integration_test.go:37]: In the scenario you described, wouldn't it be better to set the env vars rather than skip the integration tests? If anyone's changing this client, I would want them to run the integration tests.

Comment (andrewoh531) [reply client_integration_test.go:37]: Having said that, we currently don't have anywhere to store/share the API Keys securely for the client libraries amongst ourselves and future developers.

**Review by dagoof — APPROVED** [review]

File: client_integration_test.go (hunk 23 above)
Comment (jonathaningram) [reply client_integration_test.go:37]: @andrewoh531 I think it's fine to not do skip for now. Just something to revisit and see if we should do it.

File: client_integration_test.go [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+	phoneNumber := os.Getenv("SMS_RECIPIENT_NUMBER")
+	if phoneNumber == "" {
+		t.Fatal("could not set SMS_RECIPIENT_NUMBER environment variable")
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 change message to same as above

File: client_integration_test.go [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+		emailTemplateID,
+		emailAddress,
+		notify.Personalisation{
+			{"name", "John"},
+			{"colour", "pink"},
+		},
+		notify.Reference(ref),
+	)
+	if err != nil {
+		t.Fatalf("Error sending email: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+	if resp.Content.Body != wantedBody {
+		t.Errorf("got %v, want %v", resp.Content.Body, wantedBody)
+	}
+}
+
+func TestGetAllTemplates(t *testing.T) {
+	client, smsTemplateID, emailTemplateID := setup(t)
+	resp, err := client.GetAllTemplates("")
+	if err != nil {
+		t.Fatalf("Error fetching all templates: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	assertTemplateFound(t, resp, []string{emailTemplateID, smsTemplateID})
+}
+
+func TestGetSMSTemplates(t *testing.T) {
+	client, smsTemplateID, _ := setup(t)
+	resp, err := client.GetAllTemplates("sms")
+	if err != nil {
+		t.Fatalf("Error fetching sms templates: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	assertTemplateFound(t, resp, []string{smsTemplateID})
+}
+
+func TestGetEmailTemplates(t *testing.T) {
+	client, _, emailTemplateID := setup(t)
+	resp, err := client.GetAllTemplates("email")
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 29]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	assertTemplateFound(t, resp, []string{emailTemplateID})
+}
+
+func TestGetTemplateByID(t *testing.T) {
+	client, smsTemplateID, _ := setup(t)
+	resp, err := client.GetTemplateByID(smsTemplateID)
+	if err != nil {
+		t.Fatalf("Error fetching template by id: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+	if resp.Type != wantedType {
+		t.Errorf("got %v, want %v", resp.Type, wantedType)
+	}
+}
+
+func TestGetTemplateByIDAndVersion(t *testing.T) {
+	client, _, emailTemplateID := setup(t)
+	resp, err := client.GetTemplateByIDAndVersion(emailTemplateID, 1)
+	if err != nil {
+		t.Fatalf("Error fetching template version: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 31]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+func TestGenerateTemplatePreview(t *testing.T) {
+	client, _, emailTemplateID := setup(t)
+	resp, err := client.GenerateTemplatePreview(emailTemplateID, notify.Personalisation{
+		{"name", "KD"},
+		{"colour", "yellow"},
+	})
+	if err != nil {
+		t.Fatalf("Error fetching email templates: %v", err)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

File: client_integration_test.go [hunk 32]
```diff
@@ -0,0 +1,206 @@
...
+func assertTemplateFound(t *testing.T, templates notify.Templates, templateIDs []string) {
+	for _, id := range templateIDs {
+		found := false
+		for _, template := range templates {
+			if template.ID == id {
+				found = true
+			}
+		}
+		if !found {
+			t.Errorf("Template id not found: %s", id)
```
Comment (jonathaningram) [inline client_integration_test.go]: @andrewoh531 `could not`

**Review by jonathaningram — APPROVED** [review]
@andrewoh531 good work. Left a few comments, just nits on the failure descriptions.
